kuka_sunrise_control/teleop_guided_robot/src/control_system/system_manager.cpp

```diff
@@ -14,107 +14,211 @@
+      response->success = true;
+
+      if (this->get_current_state().label() == "active") {
+        std_msgs::msg::Bool activate;
+        activate.data = false;
+        manage_processing_publisher_->publish(activate);
+        if (this->deactivate().label() != "inactive") {
+          response->success = false;
+        }
+        RCLCPP_WARN(
+          get_logger(),
+          "Motion stopped externally, deactivating controls and managers");
+      } else {
+        RCLCPP_WARN(
+          get_logger(),
+          "Invalid request, system manager not active");
+      }
+    };
+  trigger_change_service_ = this->create_service<std_srvs::srv::Trigger>(
+    "system_manager/trigger_change", trigger_change_callback);
+
+  param_callback_ = this->add_on_set_parameters_callback(
+    [this](const std::vector<rclcpp::Parameter> & parameters) {
+      return getParameterHandler().onParamChange(parameters);
+    });
+
+  // If node is in unconfigured state, parameter can be always set
+  registerParameter<bool>(
+    "control_logic", false, kroshu_ros2_core::ParameterSetAccessRights {true, false,
+      false, false}, [this](bool control_logic) {
+      control_logic_ = control_logic;
+      return true;
+    });
 }
 
 rclcpp_lifecycle::node_interfaces::LifecycleNodeInterface::CallbackReturn SystemManager::
 on_configure(
-  const rclcpp_lifecycle::State & state)
+  const rclcpp_lifecycle::State &)
 {
-  (void)state;
-  if (!changeState(ROBOT_INTERFACE, lifecycle_msgs::msg::Transition::TRANSITION_CONFIGURE)) {
-    return FAILURE;
+  auto result = SUCCESS;
+  if (!changeState(
+      ROBOT_INTERFACE,
+      lifecycle_msgs::msg::Transition::TRANSITION_CONFIGURE))
+  {
+    result = FAILURE;
   }
-  if (!changeState(JOINT_CONTROLLER, lifecycle_msgs::msg::Transition::TRANSITION_CONFIGURE)) {
-    return ERROR;
+  if (result == SUCCESS && !changeState(
+      JOINT_CONTROLLER,
+      lifecycle_msgs::msg::Transition::TRANSITION_CONFIGURE))
+  {
+    result = FAILURE;
   }
-  if (!changeState(CONTROL_LOGIC, lifecycle_msgs::msg::Transition::TRANSITION_CONFIGURE)) {
-    return ERROR;
+
+  if (control_logic_ && result == SUCCESS &&
+    !changeState(CONTROL_LOGIC, lifecycle_msgs::msg::Transition::TRANSITION_CONFIGURE))
+  {
+    result = FAILURE;
   }
-  return SUCCESS;
+  if (result != SUCCESS) {
+    this->on_cleanup(get_current_state());
+  }
+
+  return result;
 }
 
 rclcpp_lifecycle::node_interfaces::LifecycleNodeInterface::CallbackReturn SystemManager::on_cleanup(
-  const rclcpp_lifecycle::State & state)
+  const rclcpp_lifecycle::State &)
 {
-  (void)state;
-  if (!changeState(ROBOT_INTERFACE, lifecycle_msgs::msg::Transition::TRANSITION_CLEANUP)) {
-    return FAILURE;
+  if (getState(ROBOT_INTERFACE).id != lifecycle_msgs::msg::State::PRIMARY_STATE_UNCONFIGURED &&
+    !changeState(
+      ROBOT_INTERFACE,
+      lifecycle_msgs::msg::Transition::TRANSITION_CLEANUP))
+  {
+    return ERROR;
```

Comment:
same here, why ERROR instead of FAILURE?

Reply:
For deactivation and cleanup, I felt it is safer to return error if transition is not successful, in these cases, the node will go to error state and get finalized or return to unconfigured.

For configure or activate, this is not the desired behaviour, it is safe to return to the original state and thus enable retrying the transition